### PR TITLE
API docs documentation

### PR DESCRIPTION
**Pull request by miguelangel-dev**
## Description
Complete API docs for `nef`. It reaches the goal of 100% code coverage.

### Documented modules:
- `nef`
- `NefModels`

Captura de pantalla 2020-03-23 a las 14 19 20

### CODE REVIEW

File: project/Component/NefModels/CarbonModel.swift [hunk 1]
```diff
@@ -2,25 +2,54 @@
 
 import Foundation
 
+/// Represents the configuration for rendering carbon snippets.
 public struct CarbonModel: Codable, Equatable {
+    /// Code to render into Carbon image.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
    /// Code to render as a Carbon image.
```

File: project/Component/NefModels/CarbonModel.swift [hunk 2]
```diff
@@ -2,25 +2,54 @@
...
     public let code: String
+    
+    /// Style to apply to the generated snippet.
     public let style: CarbonStyle
     
+    /// Initializes a `CarbonModel`
+    ///
+    /// - Parameters:
+    ///   - code: code to render into Carbon image.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
    ///   - code: code to render as a Carbon image.
```

File: project/Component/NefModels/CarbonModel.swift [hunk 3]
```diff
@@ -2,25 +2,54 @@
...
     }
 }
 
-// MARK: - Style
+/// Describes the style to apply to Carbon snippet.
 public struct CarbonStyle: Codable, Equatable {
+    /// Background color
     public let background: Color
+    
+    /// File size.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]: I think *file size* is misguiding.
```suggestion
    /// Image size.
```

File: project/Component/NefModels/CarbonModel.swift [hunk 4]
```diff
@@ -2,25 +2,54 @@
...
+    
+    /// Shows/hides nef watermark.
     public let watermark: Bool
     
+    /// Initializes a `CarbonStyle`
+    ///
+    /// - Parameters:
+    ///   - background: background color in hexadecimal.
+    ///   - theme: carbon theme.
+    ///   - size: export file size.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
    ///   - size: exported image size.
```

File: project/Component/NefModels/CarbonModel.swift [hunk 5]
```diff
@@ -2,25 +2,54 @@
...
+    /// Shows/hides lines of code.
     public let lineNumbers: Bool
+    
+    /// Shows/hides nef watermark.
     public let watermark: Bool
     
+    /// Initializes a `CarbonStyle`
+    ///
+    /// - Parameters:
+    ///   - background: background color in hexadecimal.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]: The hexadecimal detail should be hidden in the implementation of `Color`.

```suggestion
    ///   - background: background color.
```

File: project/Component/NefModels/CarbonModel.swift [hunk 6]
```diff
@@ -2,25 +2,54 @@
...
+    ///   - theme: carbon theme.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
    ///   - theme: Carbon theme.
```

File: project/Component/NefModels/CarbonModel.swift [hunk 7]
```diff
@@ -2,25 +2,54 @@
...
+    ///   - size: export file size.
+    ///   - fontType: carbon font type.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
    ///   - fontType: Carbon font type.
```

File: project/Component/NefModels/CarbonModel.swift [hunk 8]
```diff
@@ -30,70 +59,166 @@ public struct CarbonStyle: Codable, Equatable {
         self.watermark = watermark
     }
     
+    /// Represents file size.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
    /// Represents image size.
```

File: project/Component/NefModels/CarbonModel.swift [hunk 9]
```diff
@@ -30,70 +59,166 @@ public struct CarbonStyle: Codable, Equatable {
...
     public enum Size: Double, Codable, Equatable, CaseIterable {
+        /// Output file size to 1x
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
        /// Output image size to 1x
```

File: project/Component/NefModels/CarbonModel.swift [hunk 10]
```diff
@@ -30,70 +59,166 @@ public struct CarbonStyle: Codable, Equatable {
...
         case x1 = 1
+        /// Output file size to 2x
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
        /// Output image size to 2x
```

File: project/Component/NefModels/CarbonModel.swift [hunk 11]
```diff
@@ -30,70 +59,166 @@ public struct CarbonStyle: Codable, Equatable {
...
         case x2 = 2
+        /// Output file size to 3x
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
        /// Output image size to 3x
```

File: project/Component/NefModels/CarbonModel.swift [hunk 12]
```diff
@@ -30,70 +59,166 @@ public struct CarbonStyle: Codable, Equatable {
...
         case x3 = 3
+        /// Output file size to 4x
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
        /// Output image size to 4x
```

File: project/Component/NefModels/CarbonModel.swift [hunk 13]
```diff
@@ -30,70 +59,166 @@ public struct CarbonStyle: Codable, Equatable {
...
         case x4 = 4
+        /// Output file size to 5x
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
        /// Output image size to 5x
```

File: project/Component/NefModels/CarbonModel.swift [hunk 14]
```diff
@@ -30,70 +59,166 @@ public struct CarbonStyle: Codable, Equatable {
...
         case x5 = 5
     }
     
+    /// Represents Carbon theme.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
    /// Represents a Carbon theme.
```

File: project/Component/NefModels/CarbonModel.swift [hunk 15]
```diff
@@ -30,70 +59,166 @@ public struct CarbonStyle: Codable, Equatable {
...
     public enum Theme: String, Codable, Equatable, CaseIterable {
+        /// Base 16 dark
         case base16 = "base16-dark"
+        
+        /// Balckboard
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
        /// Blackboard
```

File: project/Component/NefModels/CarbonModel.swift [hunk 16]
```diff
@@ -30,70 +59,166 @@ public struct CarbonStyle: Codable, Equatable {
...
         case verminal
+        
+        /// VSCode
         case vscode
+        
+        /// Zenburn
         case zenburn
     }
     
+    /// Represents Carbon font type
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
    /// Represents a Carbon font type
```

File: project/Component/NefModels/CarbonModel.swift [hunk 17]
```diff
@@ -124,30 +253,56 @@ extension CarbonStyle.Color {
...
 }
 
-// MARK: Error
+/// Represents an error in a Carbon action.
 public struct CarbonError: Error {
+    
+    /// The snippet of code fails in the action.
     public let snippet: String
+    
+    /// Describe the reason for the failure.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
    /// Describes the reason for the failure.
```

File: project/Component/NefModels/CarbonModel.swift [hunk 18]
```diff
@@ -124,30 +253,56 @@ extension CarbonStyle.Color {
...
+    
+    /// Predefined orange `Color`.
     public static let orange = CarbonStyle.Color(r: 255, g: 159, b: 70, a: 1)
 }
 
-// MARK: Error
+/// Represents an error in a Carbon action.
 public struct CarbonError: Error {
+    
+    /// The snippet of code fails in the action.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
    /// Snippet that was being rendered when the action failed.
```

File: project/Component/NefModels/CarbonModel.swift [hunk 19]
```diff
@@ -124,30 +253,56 @@ extension CarbonStyle.Color {
...
     public let snippet: String
+    
+    /// Describe the reason for the failure.
     public let cause: CarbonError.Cause
     
+    /// Initializes a `CarbonError`
+    ///
+    /// - Parameters:
+    ///   - snippet: the snippet of code fails in the action
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
    ///   - snippet: snippet of code that was being rendered when the action failed.
```

File: project/Component/NefModels/CarbonModel.swift [hunk 20]
```diff
@@ -124,30 +253,56 @@ extension CarbonStyle.Color {
...
+    ///   - cause: the reason for the failure.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
    ///   - cause: reason for the failure.
```

File: project/Component/NefModels/CarbonModel.swift [hunk 21]
```diff
@@ -124,30 +253,56 @@ extension CarbonStyle.Color {
...
-    // MARK: Error options
+    /// Models the reason of a carbon error.
     public enum Cause: Error, CustomStringConvertible {
+        ///  Could not open Carbon web view with the current configuration.
         case notFound
+        
+        /// Could not take a snapshot.
         case invalidSnapshot
         
+        /// A textual representation of `CarbonError.Cause`.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]: I think this does not need to be documented, as its documentation is inherited from `CustomStringConvertible`.

File: project/Component/NefModels/CarbonModel.swift [hunk 22]
```diff
@@ -157,8 +312,11 @@ public struct CarbonError: Error {
     }
 }
 
-// MARK: - Constructors
 extension CarbonStyle.Size {
+    
+    /// Initializes a `CarbonStyle.Size`
+    ///
+    /// - Parameter factor: export file size.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
    /// - Parameter factor: export image size.
```

File: project/Component/NefModels/CarbonModel.swift [hunk 23]
```diff
@@ -167,6 +325,10 @@ extension CarbonStyle.Size {
 }
 
 extension CarbonStyle.Color {
+    
+    /// Initializes a `CarbonStyle.Color`
+    ///
+    /// - Parameter hex: creates a `Color` from an hexadecimal.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
    /// - Parameter hex: creates a `Color` from a hexadecimal String.
```

File: project/Component/NefModels/Console.swift [hunk 24]
```diff
@@ -16,25 +28,59 @@ public struct Step {
 }
 
 extension Step {
+    /// Defined dummy `Step`
     public static var empty: Step { .init(total: 0, partial: 0, duration: .never) }
     
+    /// Advance i-operations the step.
```
Comment (truizlop) [inline project/Component/NefModels/Console.swift]:
```suggestion
    /// Advance the step a number of increments.
```

File: project/Component/NefModels/Console.swift [hunk 25]
```diff
@@ -16,25 +28,59 @@ public struct Step {
...
+    /// - Parameter partial: number of i-operations.
```
Comment (truizlop) [inline project/Component/NefModels/Console.swift]:
```suggestion
    /// - Parameter partial: number of increments for the step.
```

File: project/Component/nef/Utils/Kleisli+API.swift [hunk 26]
```diff
@@ -5,16 +5,30 @@ import NefModels
 import Bow
 import BowEffects
 
+/// Transforms an `EnvIO<Console, nef.Error, A>` into `IO<nef.Error, A>` using a dummy console.
 public extension Kleisli where D == NefModels.Console, F == IOPartial<nef.Error> {
+    
+    /// Performs the side effects that are suspended in this IO in a synchronous manner. Resolve `D` dependency of type `Console` using a dummy console.
```
Comment (truizlop) [inline project/Component/nef/Utils/Kleisli+API.swift]:
```suggestion
    /// Performs the side effects that are suspended in this IO in a synchronous manner, resolving the dependency of type `Console` using a dummy implementation.
```

File: project/Component/nef/Utils/Kleisli+API.swift [hunk 27]
```diff
@@ -5,16 +5,30 @@ import NefModels
...
+    /// - Parameter queue: Dispatch queue used to execute the side effects. Defaults to the main queue.
+    /// - Returns: Value produced after running the suspended side effects.
+    /// - Throws: Error of type `nef.Error` that may happen during the evaluation of the side-effects.
     func unsafeRunSync(on queue: DispatchQueue = .main) throws -> A {
         try provide(MacDummyConsole()).unsafeRunSync()
     }
     
+    /// Performs the side effects that are suspended in this IO in a synchronous manner. Resolve `D` dependency of type `Console` using a dummy console.
```
Comment (truizlop) [inline project/Component/nef/Utils/Kleisli+API.swift]:
```suggestion
    /// Performs the side effects that are suspended in this IO in a synchronous manner, resolving the dependency of type `Console` using a dummy implementation.
```

File: project/Component/nef/Utils/Kleisli+API.swift [hunk 28]
```diff
@@ -5,16 +5,30 @@ import NefModels
...
+    /// - Parameter queue: Dispatch queue used to execute the side effects. Defaults to the main queue.
+    /// - Returns: Value produced after running the suspended side effects.
     func unsafeRunSyncEither(on queue: DispatchQueue = .main) -> Either<nef.Error, A> {
         provide(MacDummyConsole()).unsafeRunSyncEither()
     }
     
-    func unsafeRunAsync(with d: D, on queue: DispatchQueue = .main, _ callback: @escaping Callback<nef.Error, A>) {
+    /// Performs the side effects that are suspended in this IO in an asynchronous manner. Resolve `D` dependency of type `Console` using a dummy console.
```
Comment (truizlop) [inline project/Component/nef/Utils/Kleisli+API.swift]:
```suggestion
    /// Performs the side effects that are suspended in this IO in an asynchronous manner, resolving the dependency of type `Console` using a dummy implementation.
```

File: project/Component/nef/Utils/Kleisli+API.swift [hunk 29]
```diff
@@ -5,16 +5,30 @@ import NefModels
...
+    /// - Parameters
+    ///   - queue: Dispatch queue used to execute the side effects. Defaults to the main queue.
+    ///   - callback: A callback function to receive the results of the evaluation.
+    /// - Returns: Value produced after running the suspended side effects.
```
Comment (truizlop) [inline project/Component/nef/Utils/Kleisli+API.swift]: It does not return anything.

```suggestion
```

**Review by truizlop — CHANGES_REQUESTED** [review]

**Comment by truizlop** [comment]
A general comment: the description of each parameter should start with a capital letter, as it is the beginning of a sentence. Use that as a convention.

File: project/Component/NefModels/CarbonModel.swift (hunk 21 above)
Comment (miguelangel-dev) [reply project/Component/NefModels/CarbonModel.swift]: yep, you are right - I think so, anyway, I will check it

File: project/Component/NefModels/CarbonModel.swift [hunk 30]
```diff
@@ -2,25 +2,54 @@
...
     
+    /// Initializes a `CarbonStyle`
+    ///
+    /// - Parameters:
+    ///   - background: Background color.
+    ///   - theme: Carbon theme.
+    ///   - size: Exported image size.
+    ///   - fontType: Carbon font type.
+    ///   - lineNumbers: Shows/hides lines of code.
+    ///   - watermark:Shows/hides the watermark.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
    ///   - watermark: Shows/hides the watermark.
```

File: project/Component/NefModels/CarbonModel.swift [hunk 31]
```diff
@@ -30,70 +59,166 @@ public struct CarbonStyle: Codable, Equatable {
...
+        /// Hopscotch
         case hopscotch
+        
+        /// Lucario
         case lucario
+        
+        /// Material
         case material
+        
+        ///Monokai
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
        /// Monokai
```

File: project/Component/NefModels/CarbonModel.swift [hunk 32]
```diff
@@ -114,8 +240,10 @@ extension CarbonStyle: CustomStringConvertible {
     }
 }
 
-// MARK: Default colors
+/// Predefined Carbon `Color`.
 extension CarbonStyle.Color {
+    
+    /// A group of predefined Carbon colours.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]:
```suggestion
    /// A group of predefined Carbon colors.
```

File: project/Component/NefModels/CarbonView.swift [hunk 33]
```diff
@@ -2,12 +2,22 @@
 
 import AppKit
 
+/// Describes a Carbon loading view.
 public protocol CarbonLoadingView: NSView {
+    
+    /// Loading view should change its state to show.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonView.swift]:
```suggestion
    /// Shows the loading view.
```

File: project/Component/NefModels/CarbonView.swift [hunk 34]
```diff
@@ -2,12 +2,22 @@
...
     func show()
+    
+    /// Loading view should change its state to hidden.
```
Comment (truizlop) [inline project/Component/NefModels/CarbonView.swift]:
```suggestion
    /// Hides the loading view.
```

File: project/Component/nef/Models/Error.swift [hunk 35]
```diff
@@ -2,17 +2,32 @@
 
 import Foundation
 
+/// Models a `nef` error.
 public enum Error: Swift.Error {
+    /// Describes a `Compiler` error given detailed information.
```
Comment (truizlop) [inline project/Component/nef/Models/Error.swift]:
```suggestion
    /// Describes a `Compiler` error with detailed information.
```

File: project/Component/nef/Models/Error.swift [hunk 36]
```diff
@@ -2,17 +2,32 @@
...
     case compiler(info: String = "")
+    
+    /// Describes a `Markdown` error given detailed information.
```
Comment (truizlop) [inline project/Component/nef/Models/Error.swift]:
```suggestion
    /// Describes a `Markdown` error with detailed information.
```

File: project/Component/nef/Models/Error.swift [hunk 37]
```diff
@@ -2,17 +2,32 @@
...
     case markdown(info: String = "")
+    
+    /// Describes a `Jekyll` error given detailed information.
```
Comment (truizlop) [inline project/Component/nef/Models/Error.swift]:
```suggestion
    /// Describes a `Jekyll` error with detailed information.
```

File: project/Component/nef/Models/Error.swift [hunk 38]
```diff
@@ -2,17 +2,32 @@
...
     case jekyll(info: String = "")
+    
+    /// Describes a `Carbon` error given detailed information.
```
Comment (truizlop) [inline project/Component/nef/Models/Error.swift]:
```suggestion
    /// Describes a `Carbon` error with detailed information.
```

File: project/Component/nef/Models/Error.swift [hunk 39]
```diff
@@ -2,17 +2,32 @@
...
     case carbon(info: String = "")
+    
+    /// Describes a `Carbon` error given detailed information about the snapshot failure.
```
Comment (truizlop) [inline project/Component/nef/Models/Error.swift]:
```suggestion
    /// Describes a `Carbon` error with detailed information about the snapshot failure.
```

File: project/Component/nef/Models/Error.swift [hunk 40]
```diff
@@ -2,17 +2,32 @@
...
     case invalidSnapshot(info: String = "")
+    
+    /// Describes a `Playground` error given detailed information.
```
Comment (truizlop) [inline project/Component/nef/Models/Error.swift]:
```suggestion
    /// Describes a `Playground` error with detailed information.
```

File: project/Component/nef/Models/Error.swift [hunk 41]
```diff
@@ -2,17 +2,32 @@
...
     case playground(info: String = "")
+    
+    /// Describes a `Swift Playground` error given detailed information.
```
Comment (truizlop) [inline project/Component/nef/Models/Error.swift]:
```suggestion
    /// Describes a `Swift Playground` error with detailed information.
```

File: project/Component/nef/Utils/Kleisli+API.swift [hunk 42]
```diff
@@ -5,16 +5,29 @@ import NefModels
 import Bow
 import BowEffects
 
+/// Transforms an `EnvIO<Console, nef.Error, A>` into `IO<nef.Error, A>` using a dummy console.
```
Comment (truizlop) [inline project/Component/nef/Utils/Kleisli+API.swift:8]: I think this documentation is ignored.

**Review by truizlop — CHANGES_REQUESTED** [review]

File: project/Component/nef/Utils/Kleisli+API.swift (hunk 42 above)
Comment (miguelangel-dev) [reply project/Component/nef/Utils/Kleisli+API.swift:8]: it is optional, but if you add it, it shows as a description, on bottom of the signature:
Captura de pantalla 2020-03-23 a las 17 19 02

**Review by truizlop — APPROVED** [review]

**Pull request merged**